### PR TITLE
Fix battery discharge showing on ABUPS Fuel Fired Power Generation

### LINKED ISSUE #4419: ABUPS Electric Loads Satisfied value for 'Fuel-Fired Power Generation"  is incorrect when using PV and battery storage
Helpdesk Ticket 9123

Posted on: 12 March 2014 10:20 PM
I ran without the BCVTB and confirmed that the value reported for "Fuel-Fired Power Generation" is incorrect. It is the sum of these two output variables, which is clearly wrong.

 BATERIJA:Electric Storage Production Decrement Energy
 BATERIJA:Electric Storage Discharge Energy

 In order to track the performance of your PV system and battery storage, you will need to use the relevant output:variable results. You can access these also using Output:Table:Monthly (one for the storage, one for the pv generators, and one for the distribution center) if you want results in the table output.

 Thank you for calling this to our attention.
 Mike

 > Hello!
 > 
 > I've attached used IDF and weather file. Please let me know, what did
 > you
 > find... For simulation I'm using BCVBT.

2014-03-06 19:12 GMT+01:00 EnergyPlus Support <energyplus-support@gard.com>:
 > My apologies for the slow response. Please reply and attach your idf file
 > and tell us what epw file you are using.
 >
 > To answer your first question, energy drawn from the battery is not
 > represented as Fuel-Fired Power Generation.
 >
 > Mike

Posted on: 18 February 2014 08:32 AM
Hello!

 I have a simple model with added photovoltaic panels and KiBaM battery
 model to store power.
 Everything seems to work ok, but I have one question regarding outputs.
 Is it possible that energy that is drawn from battery is represented
 as "Fuel-Fired
 Power Generation" in final summary report <modelname>Table.csv
 /<modelname>Table.html under Electric Loads Satisfied?

 I noticed that PV output power is more or less the same if I use some kind
 of electrical storage or not. Difference is shown when I add battery or
 simple storage. It somehow adds almost all the power, that is needed for
 operating this simple building and final report is strange(difference
 almost doubles as if there is double the amount of produced electric power
 EVEN if battery is empty at the beginning of simulation).
 Is this bug in simulation software or is there something that I'm missing.
 Is my PV grid producing more power than it could be fed into the grid and
 difference shows only when there is another "consumer" - battery?

 My model is using PhotovoltaicPerformanceSimple, with 11% efficiency, 0.9
 fraction of roof surface, which roughly measures 7m x 2.5m.
 Electric load center:generators uses 1 PV generator, rated electric power
 output 3000W.
 Battery is modeled with same specification that is available in
 InputOutputreference.pdf - 10 modules in parallel, 10 in series...

 If you have any idea what could be the cause of this, to my understanding a
 bit odd situation, please let me know.

 Kind regards,
 Matej Krebelj

User files
ContainerWithBattery.idf
Jul_Aug_Sep_Okt_Nov_2013CSTHourly.epw

### PR DESCRIPTION
**Pull request by JasonGlazer**
Fixes #4419

### CODE REVIEW

**Comment by EnergyArchmage** [comment]
I do not think this correct.  The idea here is that battery storage introduces a loss of produced energy because the round trip through storage isn't 100% efficient.  This change will do away with that accounting of storage losses and I think that is a mistake.  

For this issue you could just change from "Cogeneration" to "Photovoltaics"  but the problem has always been that the battery could be used with either pv or fuel-fired production or a wind turbine.  
While it is true that this metering is showing up in an odd place, that is only because there isn't a better meter nor a better place in the in ABUPS to put this.   I suppose it might be possible to detect what sort of equipment is producing the electricity and shift the setup call between Cogeneration, Photovoltaics, or WindTurbine depending on what is likely to be the source.   

Or you could just add to the explanation of these output variables in the I/O ref to point out that the battery losses appear on the Cogeneration meter by design because of these issues.  

The full fix, which is a lot of work, would be to add another submeter called "Storage" that is analagous to Photovoltaics, WindTurbine, HTGeothermal, etc and then also another new row to the Electric Loads Satisfied table and associated accumulation to the tabular reporting system.

**Comment by JasonGlazer** [comment]
@EnergyArchmage @mjwitte I changed the approach to match what I think was the intent that Brent suggested. The ABUPS table still looks a little weird using the defect file from:

https://github.com/NREL/EnergyPlusDevSupport/tree/master/DefectFiles/4000s/4419

and is shown below:

capture

File: src/EnergyPlus/ManageElectricPower.cc [hunk 1]
```diff
@@ -1339,10 +1339,12 @@ namespace ManageElectricPower {
...
+				SetupOutputVariable( "Electric Storage Charge Energy [J]", ElecStorage( StorNum ).StoredEnergy, "System", "Sum", ElecStorage( StorNum ).Name, _, "Electricity", "Cogeneration", "Batteries", "Plant" );
+				SetupOutputVariable( "Electric Storage Production Decrement Energy [J]", ElecStorage( StorNum ).DecrementedEnergyStored, "System", "Sum", ElecStorage( StorNum ).Name );
 				SetupOutputVariable( "Electric Storage Discharge Power [W]", ElecStorage( StorNum ).DrawnPower, "System", "Average", ElecStorage( StorNum ).Name );
-				SetupOutputVariable( "Electric Storage Discharge Energy [J]", ElecStorage( StorNum ).DrawnEnergy, "System", "Sum", ElecStorage( StorNum ).Name, _, "ElectricityProduced", "COGENERATION", _, "Plant" );
+				SetupOutputVariable( "Electric Storage Discharge Energy [J]", ElecStorage( StorNum ).DrawnEnergy, "System", "Sum", ElecStorage( StorNum ).Name);
+				ElecStorage( StorNum ).DecrementDrawnEnergy = -1.0 * ElecStorage( StorNum ).DrawnEnergy;
```
Comment (mjwitte) [inline src/EnergyPlus/ManageElectricPower.cc]: @JasonGlazer I believe this line is misplaced.  It needs to be in the same place that .DrawnEnergy is being set.

Comment (JasonGlazer) [inline src/EnergyPlus/ManageElectricPower.cc]: Your right. Thanks.

**Comment by EnergyArchmage** [comment]
I think all the code related to DecrementedEnergyStored could be cleaned out and replaced with similiar code for DecrementedEnergyDrawn. 

Array1.Project.MArray.hh may need code analgous to this but I do not understand this part of the code.
    inline MA1c_double DecrementedEnergyStored() const { return ma( &T::DecrementedEnergyStored ); }
    inline MA1_double DecrementedEnergyStored() { return ma( &T::DecrementedEnergyStored ); }

The negative value that looks odd will occur when the state of charge drops over the simulation -- when the battery charging is not keeping up and the battery is full at the start and empty at the end.  This is confusing when the information does not appear along with the Net Decrease in On-Site Storage in the Electric Loads Satisfied table.  Perhaps you could do a one-time adjustment to ABUPS data at the end of the run period to adjust the end use to account for Net Decrease in On-Site Storage.  It may double count if not done carefully.

**Comment by JasonGlazer** [comment]
I just fixed the location of the DecrementDrawnEnergy calcs and the results are really:

capture1
capture2

**Comment by JasonGlazer** [comment]
I am not sure I understand the inline code changes either but it seems to work without them.

**Comment by mjwitte** [comment]
A negative value in the main ABUPS table doesn't make sense here.  The storage isn't creating energy.

**Comment by EnergyArchmage** [comment]
Okay, I checked this branch out, built it and ran the example files ShopWithPVandStorage and ShopWithPVandBattery.  ShopWithPVandBattery and the defect file use the newer kinetic lead acid model and they both come out with strange negatives there.  ShopWithPVandStorage works as expected. 

It looks like the kinetic battery model got  implemented with a different sign convention than the simple storage bucket.  So the metering is/was all backwards for that one and this is a serious issue of incorrect metering in addition to the issue of how reporting is organized.

**Comment by JasonGlazer** [comment]
Thanks Brent. Here are the tables from the ABUPS report:

capture3
capture4

The sign for generators on the end use table looks more reasonable.

**Comment by mjwitte** [comment]
But the magnitude seems terribly high - more power lost in the storage than produced by the PV?  How can that be if the only source that can charge the battery is the on-site PV generation?

**Comment by EnergyArchmage** [comment]
Yeah, that is not good.  ShopWithPVandBattery does not have the same problem.  Can you refresh the defect file you are using?

**Comment by JasonGlazer** [comment]
I am not sure what you mean by refresh. The defect files are at:

https://github.com/NREL/EnergyPlusDevSupport/tree/master/DefectFiles/4000s/4419

**Comment by EnergyArchmage** [comment]
oh sorry, I see a v8.4 file now, _I_ needed to refresh.  I see now that the discharge power didn't need its sign flipped and reverted that change here https://github.com/NREL/EnergyPlus/pull/5263#.   There are other issues with the kinetic battery model that would need to get fixed before the results will be good.  I think you want to restrict this to just the original tabular reporting issue and suggest to use example file ShopWithPVandStorage.idf as the defect file.

**Comment by mjwitte** [comment]
Merged in develop to get clean test results.

**Comment by EnergyArchmage** [comment]
Note that the test failures here are expected.  End use subcategories were added for batteries, inverters, and transformers and they show up as string diffs in html reports and changes in the table sizes.

**Comment by mjwitte** [comment]
@EnergyArchmage But at least one case has html file changes that I can't explain.  Look at eplustbl.htm.absdiff.htm [here](http://energyplus.s3-website-us-east-1.amazonaws.com/regressions/2015-10/621b8629c56175639755f366ce858e44b175086c-bbd7b7528e5429ffa7439b5ba7b0e101fb387610/ShopWithSimplePVT/Win64-Windows-7-VisualStudio-12/).  Why are they small changes in time setpoint not met and time not comfortable?

**Comment by EnergyArchmage** [comment]
The reason is that the transformer is pulling the on-site power generation from  "Cogeneration:ElectricityProduced", which has changed, and the transformer has thermal losses to the zone.  This shows a very good reason why these losses do need to be moved over to the be end uses. 

But...  I noticed that transformer losses also get put on the the fuel-fired generator (nastly little feedback loop there I suspect). 
                SetupOutputVariable( "Transformer Cogeneration Electric Loss Energy [J]", Transformer( TransfNum ).ElecProducedCoGen, "System", "Sum", Transformer( TransfNum ).Name, _, "ElectricityProduced", "COGENERATION", "Transformer", 

Anyway so the transformer losses also need to be moved as they should follow the battery's pattern.  I will push up some code for that shortly.

**Comment by mjwitte** [comment]
@EnergyArchmage Be aware of #4622 which added the transformers to an end-use so that the total electricity in the ABUPS report would match the sum of the rows.

**Comment by EnergyArchmage** [comment]
AH shoot, it isn't the transformer, there isn't even one.  It is a radically different amount of charging and discharging in the battery.  This whole approach of metering as an end use is not working well at all.  This has to be rejected entirely.   It is circular, the amount of power left over for storage cannot depend on the amount of power that got stored.  

(It may not be such a problem for the utility-feed version of the transformer in PR 4622). 

Changing the end uses affects the Facility Total Electric Demand Power which is a major input into the controlling of charging and discharging.   Here is plot from ShopWithSimplePVT comparing this variable between this PR and develop.

facility elec demand

Sorry, I go back to my original suggestion for the proper fix  "to add another submeter called "Storage" that is analagous to Photovoltaics, WindTurbine, HTGeothermal, etc and then also another new row to the Electric Loads Satisfied table and associated accumulation to the tabular reporting system."

**Comment by mjwitte** [comment]
Closing this, replaced by #5300

**Pull request closed without merging**